arangod/Aql/HashedCollectExecutor.cpp

```diff
@@ -0,0 +1,299 @@
+    output.cloneValueInto(_infos.getGroupRegisters()[i++].first,
+                          _lastInitializedInputRow, key);
+    const_cast<AqlValue*>(&key)->erase();  // to prevent double-freeing later
+  }
+
+  if (!_infos.getCount()) {
+    TRI_ASSERT(_currentGroup->second->size() == _infos.getAggregatedRegisters().size());
+    size_t j = 0;
+    for (auto const& r : *(_currentGroup->second)) {
+      output.cloneValueInto(_infos.getAggregatedRegisters()[j++].first,
+                            _lastInitializedInputRow, r->stealValue());
+    }
+  } else {
+    // set group count in result register
+    TRI_ASSERT(!_currentGroup->second->empty());
+    output.cloneValueInto(_infos.getCollectRegister(), _lastInitializedInputRow,
+                          _currentGroup->second->back()->stealValue());
+  }
+}
+
+ExecutionState HashedCollectExecutor::init() {
+  TRI_ASSERT(!_isInitialized);
+
+  // fetch & consume all input
+  while (_upstreamState != ExecutionState::DONE) {
+    InputAqlItemRow input = InputAqlItemRow{CreateInvalidInputRowHint{}};
+    std::tie(_upstreamState, input) = _fetcher.fetchRow();
+
+    if (_upstreamState == ExecutionState::WAITING) {
+      TRI_ASSERT(!input.isInitialized());
+      return ExecutionState::WAITING;
+    }
+
+    // !input.isInitialized() => _upstreamState == ExecutionState::DONE
+    TRI_ASSERT(input.isInitialized() || _upstreamState == ExecutionState::DONE);
+
+    // needed to remember the last valid input aql item row
+    // NOTE: this might impact the performance
+    if (input.isInitialized()) {
+      _lastInitializedInputRow = input;
+
+      consumeInputRow(input);
+    }
+  }
+
+  // initialize group iterator for output
+  if (!_producingOutput) {
+    _producingOutput = true;
+    // setup the iterator
+    _currentGroup = _allGroups.begin();
+  }
+
+  return ExecutionState::DONE;
+}
+
+std::pair<ExecutionState, NoStats> HashedCollectExecutor::produceRow(OutputAqlItemRow& output) {
+  TRI_IF_FAILURE("HashedCollectExecutor::produceRow") {
+    THROW_ARANGO_EXCEPTION(TRI_ERROR_DEBUG);
+  }
+
+  if (!_isInitialized) {
+    // fetch & consume all input and initialize output cursor
+    ExecutionState state = init();
+    if (state == ExecutionState::WAITING) {
+      return {state, NoStats{}};
+    }
+    TRI_ASSERT(state == ExecutionState::DONE);
+    _isInitialized = true;
+  }
+
+  // produce output
+  if (_currentGroup != _allGroups.end()) {
+    writeCurrentGroupToOutput(output);
+    _currentGroup++;
+  }
+
+  ExecutionState state = _currentGroup != _allGroups.end() ? ExecutionState::HASMORE
+                                                           : ExecutionState::DONE;
+
+  return {state, NoStats{}};
+}
+
+// if no group exists for the current row yet, this builds a new group.
+std::pair<std::unique_ptr<HashedCollectExecutor::AggregateValuesType>, std::vector<AqlValue>>
+HashedCollectExecutor::buildNewGroup(InputAqlItemRow& input, size_t n) {
+  std::vector<AqlValue> group;
```

Comment:
Please use `GroupKeyType` instead of `std::vector<AqlValue>` for consistency (my fault for not introducing it all over the place, sorry). Also in the return value.